input_output.ipynb

```diff
@@ -0,0 +1,919 @@
+    "Earlier, we mentioned the connection between bytes and strings in python. To learn more about this, let's briefly review the unicode UTF-8 standard.\n",
+    "\n",
+    "The unicode UTF-8  standard is a system to represent texts; in unicode  every character is assigned a number (the *codepoint*) that gives the location of this symbol in the table of all symbols. Particularly, the unicode UTF-8 standard stores each character as 1 to four bytes (each byte is a group of 8 bits). \n",
+    "\n",
+    "Let's see this more clearly with an example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for i in range(0, 8**3):\n",
+    "    bt = i.to_bytes(4)\n",
+    "    print(f\"bytes: {bt}, integer: {i}, string: {chr(i)}\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "The `chr` function takes an integer and returns the corresponding unicode character."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Converting bytes to text \n",
+    "\n",
+    "If we receive a `bytes` object and we want to transform it to text, we can use the `int.from_bytes` method and the`chr` function:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "message = \"Ciao\"\n",
+    "message_secret = bytes(message, \"utf-8\")\n",
+    "[print(f\"The `uft8` codepoint  is = {enc}, the bytes representation = {enc.to_bytes(4)}, the representation is {chr(enc)}\") for plain, enc in zip(message, message_secret)]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Reading/Writing CSV files\n",
+    "If you ever worked with tabular data, you surely encountered CSV (comma separated values) files. These files are used to store table in text format row by row: each row is separated by a new line and the columns inside a row are separated by commas `,` or by semicolumns `;`. The first line in the file usually contains the header giving the names of the columns:\n",
+    "\n",
+    "```\n",
+    "first_column,second_column\n",
+    "1,2\n",
+    "2,3\n",
+    "```\n",
+    "\n",
+    "As CSV is a very common to exchange tabular data such as statistics and time series, the python standard library offers facilities to read and write CSV files[^3] through the [csv](https://docs.python.org/3/library/csv.html) module. Despite this, today most people prefer using [pandas](https://pandas.pydata.org/) or [polars](https://www.pola.rs/) to manipulate tabular data because they offer more conveinence and faster handling of large datasets. These packages are outside of the scope of this tutorial and will not be handled here.\n",
```

Comment:
I didn't find a reference to `[^3] further in the text. Btw, `[^2] doesn't exist.

```suggestion
    "As CSV is very common to exchange tabular data such as statistics and time series, the Python standard library offers facilities to read and write CSV files[^3] through the [csv](https://docs.python.org/3/library/csv.html) module. Despite this, today most people prefer using [pandas](https://pandas.pydata.org/) or [polars](https://www.pola.rs/) to manipulate tabular data because they offer more convenience and faster handling of large datasets. These packages are outside of the scope of this tutorial and will not be handled here.\n",
```